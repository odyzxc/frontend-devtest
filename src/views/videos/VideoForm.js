import React from "react";
import { Button, Form } from "react-bootstrap";
import { useFormik } from "formik";
import * as Yup from "yup";

import { strings } from "common/constants";

const Input = ({ errors, touched, formik, formKey, name }) => (
  <Form.Group>
    <Form.Label htmlFor={formKey}>{name}</Form.Label>
    <Form.Control
      id={formKey}
      name={formKey}
      type="text"
      onChange={formik.handleChange}
      value={formik.values[formKey]}
      className={formik.errors[formKey] && formik.touched[formKey] && "error"}
    />
    {formik.errors[formKey] && formik.touched[formKey] && (
      <div className="input-feedback">{formik.errors[formKey]}</div>
    )}
  </Form.Group>
);

const VideoForm = ({ onSubmitSuccess, initialValues, onSubmit }) => {
  const formik = useFormik({
    initialValues: initialValues,
    onSubmit: (values, { resetForm }) => {
      onSubmit(values);
      resetForm();
      onSubmitSuccess(); // disclaimer - in a production grade app I would use some async action middleware (eg. redux-observable or redux-thunk) and fire this after receiving 200 OK after sending form data to backend
    },
    validationSchema: Yup.object().shape({
      title: Yup.string().required(strings.errorMessage.required),
      videoSnapshotThumbnailUrl: Yup.string().url(strings.errorMessage.notCorrectUrl).required(strings.errorMessage.required),
      videoSnapshotUrl: Yup.string().url(strings.errorMessage.notCorrectUrl).required(strings.errorMessage.required),
      description: Yup.string()
    })
  });
  return (
    <Form onSubmit={formik.handleSubmit}>
      <Input formKey="title" formik={formik} name="Title" />
      <Input
        formKey="videoSnapshotThumbnailUrl"
        formik={formik}
        name="Video Snapshot Thumbnail Url"
      />
      <Input
        formKey="videoSnapshotUrl"
        formik={formik}
        name="Video Snapshot Url"
      />
      <Input formKey="description" formik={formik} name="Description" />
      <Button type="submit">Submit</Button>
    </Form>
  );
};

export default VideoForm;
