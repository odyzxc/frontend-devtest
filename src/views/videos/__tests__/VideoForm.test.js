import React from "react";
import renderer from "react-test-renderer";

import VideoForm from "views/videos/VideoForm";


describe("VideoForm", () => {
  it("Should render", () => {
    const initialValues = {
      id: 1,
      title: "",
      videoSnapshotThumbnailUrl: "",
      videoSnapshotUrl: "",
      description: ""
    };
    const tree = renderer
      .create(<VideoForm initialValues={initialValues} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
