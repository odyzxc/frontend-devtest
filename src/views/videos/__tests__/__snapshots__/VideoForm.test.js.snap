// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoForm Should render 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="title"
    >
      Title
    </label>
    <input
      className="form-control"
      id="title"
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="videoSnapshotThumbnailUrl"
    >
      Video Snapshot Thumbnail Url
    </label>
    <input
      className="form-control"
      id="videoSnapshotThumbnailUrl"
      name="videoSnapshotThumbnailUrl"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="videoSnapshotUrl"
    >
      Video Snapshot Url
    </label>
    <input
      className="form-control"
      id="videoSnapshotUrl"
      name="videoSnapshotUrl"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="description"
    >
      Description
    </label>
    <input
      className="form-control"
      id="description"
      name="description"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
